design/trait-separated-from-underlying-workload-should-patch-back.md

```diff
@@ -0,0 +1,291 @@
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: example-component
+      traits:
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: Canary
+            spec:
+              canaryNumber: 1
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: NodeSelector
+            spec:
+              disktype: ssd
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: SecurityContext
+            spec:
+              runAsNonRoot: true
+```
+
+After AppConfig deployed, NodeSelector and SecurityContext trait will work in a random order.
+Assuming the controller of NodeSelector trait will work first, then it will patch and update the underlying Deployment.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      nodeSelector:
+        disktype: ssd
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+``` 
+
+This will cause first round re-create of pods, and please notice it won't trigger the canary trait to work.
+The App will become unstable at this point.
+
+At the same time, the controller of SecurityContext trait will work and update the deployment again.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      securityContext:
+        runAsNonRoot: true
+      nodeSelector:
+        disktype: ssd
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+``` 
+
+This will cause a second round of re-create. Again, the canary trait won't work.
+
+There're no doubt too many risks here in this workflow. We can conclude into two problems:
+
+1. Pods can't be re-created several times, all changes of trait in one deploy can only affect pod once.
+2. Changes of trait which will affect the underlying workload must trigger new revision and make the canary trait work.
+
+In this proposal, we're going to fix these two problems.
+
+## Goals
+
+1. make sure all changes of workload and trait only will affect underlying workload once.
+2. make sure underlying workload change will trigger revision create and trigger rollout/canary trait to work.
+
+## Proposal
+
+The overall idea of this proposal is to merge all traits and patch on the workload only once, and make this action
+trigger component revision generated automatically.
+
+To make it more clear, let's define what kind of trait we are talking with first. Let's call them `patch trait`
+for convenience. A trait will be `patch trait` only if it's change will **affect fields of underlying workload**
```

Comment:
`fields` here means `spec.template.spec`?

Reply:
Any fields of workload, `spec.template.spec` leads to the second condition "MUST trigger new revision to do rollout"